### PR TITLE
Simplify geometry types

### PR DESCRIPTION
**Pull request by wmww**
For ABI stability reasons, we previously had the int geometry types inherit from the generic ones. Making the same code work for both these inherited types and the generic ones added all sorts of complexity and a few papercuts. Now that we are breaking the mircore ABI all of that can be removed. The one downside is that the forward declarations of geometry types changed and became more verbose, so I switched everything that was forward declaring them to use the `geometry/forward.h` header. I also consolidated the headers.

### CODE REVIEW

**Comment by wmww** [comment]
It's giving me this. Is there any way to apply these changes as a patch instead of copying them in manually or setting up the specific environment needed to generate the symbols on my system (release build on debian IIRC?).

I tried a few ways of coping it into a patch file but no luck.
```
dpkg-gensymbols: error: some new symbols appeared in the symbols file: see diff output below
dpkg-gensymbols: warning: debian/libmircore2/DEBIAN/symbols doesn't match completely debian/libmircore2.symbols
--- debian/libmircore2.symbols (libmircore2_2.8.0_amd64)
+++ dpkg-gensymbolsAANi93	2022-06-14 22:55:51.419706593 +0000
@@ -1,21 +1,26 @@
 libmircore.so.2 libmircore2 #MINVER#
- MIR_CORE_2.9@MIR_CORE_2.9 2.9.0
+ MIR_CORE_2.9@MIR_CORE_2.9 2.8.0
+ _ZN3mir8geometry10Rectangles3addERKNS0_7generic9RectangleIiEE@MIR_CORE_2.9 2.8.0
+ _ZN3mir8geometry10Rectangles6removeERKNS0_7generic9RectangleIiEE@MIR_CORE_2.9 2.8.0
+ _ZN3mir8geometry10RectanglesC1ERKSt16initializer_listINS0_7generic9RectangleIiEEE@MIR_CORE_2.9 2.8.0
+ _ZN3mir8geometry10RectanglesC2ERKSt16initializer_listINS0_7generic9RectangleIiEEE@MIR_CORE_2.9 2.8.0
+ _ZNK3mir8geometry10Rectangles7confineERNS0_7generic5PointIiEE@MIR_CORE_2.9 2.8.0
  (c++|arch-bits=32)"mir::AnonymousShmFile::AnonymousShmFile(unsigned int)@MIR_CORE_2.9" 2.9.0
- (c++|arch-bits=64)"mir::AnonymousShmFile::AnonymousShmFile(unsigned long)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::AnonymousShmFile::base_ptr() const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::AnonymousShmFile::fd() const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::AnonymousShmFile::~AnonymousShmFile()@MIR_CORE_2.9" 2.9.0
- (c++)"mir::Fd::Fd()@MIR_CORE_2.9" 2.9.0
- (c++)"mir::Fd::Fd(int)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::Fd::Fd(mir::Fd&&)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::Fd::Fd(mir::IntOwnedFd)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::Fd::operator int() const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::Fd::operator=(mir::Fd)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::ShmFile::ShmFile()@MIR_CORE_2.9" 2.9.0
- (c++)"mir::ShmFile::~ShmFile()@MIR_CORE_2.9" 2.9.0
- (c++)"mir::fatal_error@MIR_CORE_2.9" 2.9.0
- (c++)"mir::fatal_error_abort(char const*, ...)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::fatal_error_except(char const*, ...)@MIR_CORE_2.9" 2.9.0
+ (c++|arch-bits=64)"mir::AnonymousShmFile::AnonymousShmFile(unsigned long)@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::AnonymousShmFile::base_ptr() const@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::AnonymousShmFile::fd() const@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::AnonymousShmFile::~AnonymousShmFile()@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::Fd::Fd()@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::Fd::Fd(int)@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::Fd::Fd(mir::Fd&&)@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::Fd::Fd(mir::IntOwnedFd)@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::Fd::operator int() const@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::Fd::operator=(mir::Fd)@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::ShmFile::ShmFile()@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::ShmFile::~ShmFile()@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::fatal_error@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::fatal_error_abort(char const*, ...)@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::fatal_error_except(char const*, ...)@MIR_CORE_2.9" 2.8.0
  (c++)"mir::geometry::Displacement::Displacement<int, int>(int&&, int&&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Displacement::Displacement<mir::geometry::detail::IntWrapper<mir::geometry::DeltaXTag>, mir::geometry::detail::IntWrapper<mir::geometry::DeltaYTag> >(mir::geometry::detail::IntWrapper<mir::geometry::DeltaXTag>&&, mir::geometry::detail::IntWrapper<mir::geometry::DeltaYTag>&&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Displacement::operator=(mir::geometry::Displacement const&)@MIR_CORE_2.9" 2.9.0
@@ -29,26 +34,26 @@
  (c++)"mir::geometry::Rectangle::Rectangle(mir::geometry::Point const&, mir::geometry::Size const&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Rectangle::Rectangle(mir::geometry::Rectangle const&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Rectangle::Rectangle(mir::geometry::Rectangle&&)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::Rectangles()@MIR_CORE_2.9" 2.9.0
+ (c++)"mir::geometry::Rectangles::Rectangles()@MIR_CORE_2.9" 2.8.0
  (c++)"mir::geometry::Rectangles::Rectangles(std::initializer_list<mir::geometry::Rectangle> const&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Rectangles::add(mir::geometry::Rectangle const&)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::begin() const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::bounding_rectangle() const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::clear()@MIR_CORE_2.9" 2.9.0
+ (c++)"mir::geometry::Rectangles::begin() const@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::geometry::Rectangles::bounding_rectangle() const@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::geometry::Rectangles::clear()@MIR_CORE_2.9" 2.8.0
  (c++)"mir::geometry::Rectangles::confine(mir::geometry::Point&) const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::end() const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::operator!=(mir::geometry::Rectangles const&) const@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::operator==(mir::geometry::Rectangles const&) const@MIR_CORE_2.9" 2.9.0
+ (c++)"mir::geometry::Rectangles::end() const@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::geometry::Rectangles::operator!=(mir::geometry::Rectangles const&) const@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::geometry::Rectangles::operator==(mir::geometry::Rectangles const&) const@MIR_CORE_2.9" 2.8.0
  (c++)"mir::geometry::Rectangles::remove(mir::geometry::Rectangle const&)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::Rectangles::size() const@MIR_CORE_2.9" 2.9.0
+ (c++)"mir::geometry::Rectangles::size() const@MIR_CORE_2.9" 2.8.0
  (c++)"mir::geometry::Size::Size()@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Size::Size(mir::geometry::Size const&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Size::Size(mir::geometry::Size&&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Size::Size<int, int>(int&&, int&&)@MIR_CORE_2.9" 2.9.0
  (c++)"mir::geometry::Size::operator=(mir::geometry::Size&&)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::geometry::operator<<(std::basic_ostream<char, std::char_traits<char> >&, mir::geometry::Rectangles const&)@MIR_CORE_2.9" 2.9.0
- (c++)"mir::mir_depth_layer_get_index(MirDepthLayer)@MIR_CORE_2.9" 2.9.0
- (c++)"typeinfo for mir::AnonymousShmFile@MIR_CORE_2.9" 2.9.0
- (c++)"typeinfo for mir::ShmFile@MIR_CORE_2.9" 2.9.0
- (c++)"vtable for mir::AnonymousShmFile@MIR_CORE_2.9" 2.9.0
- (c++)"vtable for mir::ShmFile@MIR_CORE_2.9" 2.9.0
+ (c++)"mir::geometry::operator<<(std::basic_ostream<char, std::char_traits<char> >&, mir::geometry::Rectangles const&)@MIR_CORE_2.9" 2.8.0
+ (c++)"mir::mir_depth_layer_get_index(MirDepthLayer)@MIR_CORE_2.9" 2.8.0
+ (c++)"typeinfo for mir::AnonymousShmFile@MIR_CORE_2.9" 2.8.0
+ (c++)"typeinfo for mir::ShmFile@MIR_CORE_2.9" 2.8.0
+ (c++)"vtable for mir::AnonymousShmFile@MIR_CORE_2.9" 2.8.0
+ (c++)"vtable for mir::ShmFile@MIR_CORE_2.9" 2.8.0
dpkg-gensymbols: error: some new symbols appeared in the symbols file: see diff output below
dpkg-gensymbols: error: some symbols or patterns disappeared in the symbols file: see diff output below
dpkg-gensymbols: warning: debian/libmiroil1/DEBIAN/symbols doesn't match completely debian/libmiroil1.symbols
--- debian/libmiroil1.symbols (libmiroil1_2.8.0_amd64)
+++ dpkg-gensymbolsp9XDbo	2022-06-14 22:55:51.683725903 +0000
@@ -1,13 +1,15 @@
 libmiroil.so.1 libmiroil1 #MINVER#
  MIROIL_1.0@MIROIL_1.0 2.7.0
+ _ZN6miroil22DisplayListenerWrapper11add_displayERKN3mir8geometry7generic9RectangleIiEE@MIROIL_1.0 2.8.0
+ _ZN6miroil22DisplayListenerWrapper14remove_displayERKN3mir8geometry7generic9RectangleIiEE@MIROIL_1.0 2.8.0
  (c++)"miroil::Compositor::~Compositor()@MIROIL_1.0" 2.7.0
  (c++)"miroil::DisplayConfigurationControllerWrapper::DisplayConfigurationControllerWrapper(std::shared_ptr<mir::shell::DisplayConfigurationController> const&)@MIROIL_1.0" 2.7.0
  (c++)"miroil::DisplayConfigurationControllerWrapper::set_base_configuration(std::shared_ptr<mir::graphics::DisplayConfiguration> const&)@MIROIL_1.0" 2.7.0
  (c++)"miroil::DisplayConfigurationPolicy::DisplayConfigurationPolicy()@MIROIL_1.0" 2.7.0
  (c++)"miroil::DisplayConfigurationPolicy::~DisplayConfigurationPolicy()@MIROIL_1.0" 2.7.0
  (c++)"miroil::DisplayListenerWrapper::DisplayListenerWrapper(std::shared_ptr<mir::compositor::DisplayListener> const&)@MIROIL_1.0" 2.7.0
- (c++)"miroil::DisplayListenerWrapper::add_display(mir::geometry::Rectangle const&)@MIROIL_1.0" 2.7.0
- (c++)"miroil::DisplayListenerWrapper::remove_display(mir::geometry::Rectangle const&)@MIROIL_1.0" 2.7.0
+#MISSING: 2.8.0# (c++)"miroil::DisplayListenerWrapper::add_display(mir::geometry::Rectangle const&)@MIROIL_1.0" 2.7.0
+#MISSING: 2.8.0# (c++)"miroil::DisplayListenerWrapper::remove_display(mir::geometry::Rectangle const&)@MIROIL_1.0" 2.7.0
  (c++)"miroil::DisplayListenerWrapper::~DisplayListenerWrapper()@MIROIL_1.0" 2.7.0
  (c++)"miroil::Edid::Descriptor::string_value[abi:cxx11]() const@MIROIL_1.0" 2.7.0
  (c++)"miroil::Edid::parse_data(std::vector<unsigned char, std::allocator<unsigned char> > const&)@MIROIL_1.0" 2.7.0
```

**Comment by wmww** [comment]
For some reason the 3rd and 4th lines of `debian/libmircore2.symbols` were swapped in the file the diff was based on compared to the repo. Swapping these two lines allowed it to apply:
```
(c++|arch-bits=32)"mir::AnonymousShmFile::AnonymousShmFile(unsigned int)@MIR_CORE_2.9" 2.9.0
```
and
```
(c++|arch-bits=64)"mir::AnonymousShmFile::AnonymousShmFile(unsigned long)@MIR_CORE_2.9" 2.9.0
```

**Comment by wmww** [comment]
It now seems to be complaining about libmiral symbols. Are we breaking the miral ABI? Miral uses the geometry types, so how can we break ABI on them if we're not braking miral?

**Comment by AlanGriffiths** [comment]
> It now seems to be complaining about libmiral symbols. Are we breaking the miral ABI? Miral uses the geometry types, so how can we break ABI on them if we're not braking miral?

This does break the miral ABI. I guess we should first to discuss whether we want that. (I don't have a problem with it)

File: debian/libmircore2.symbols [hunk 1]
```diff
@@ -1,21 +1,26 @@
 libmircore.so.2 libmircore2 #MINVER#
- MIR_CORE_2.9@MIR_CORE_2.9 2.9.0
- (c++|arch-bits=64)"mir::AnonymousShmFile::AnonymousShmFile(unsigned long)@MIR_CORE_2.9" 2.9.0
+ MIR_CORE_2.9@MIR_CORE_2.9 2.8.0
+ _ZN3mir8geometry10Rectangles3addERKNS0_7generic9RectangleIiEE@MIR_CORE_2.9 2.8.0
```
[the remark this answers is not in the document]
Comment (RAOF) [reply debian/libmircore2.symbols:3]: Oh. To be clear, I do *not* favour the mangled names. There are arguments in favour of them - notably, they check that we export *only* what we expect to export, and there are cases where the mangled name changes but the unmangled name does not - but I think the costs will outweigh them (*particularly* if we get proper ABI-stability testing happening). The costs largely being that name-mangling has the scope for per-architecture weirdness, so our symbols files will get more complicated and we won't be able to write them without having a failed build on the Launchpad infrastructure first.

[the remark this answers is not in the document]
Comment (wmww) [reply debian/libmircore2.symbols:3]: I just applied the diff CI gave me. I do not know how I would choose mangled or unmangled names.

**Review by AlanGriffiths — APPROVED** [review]
Seems to be viable now I've re-hacked the symbols.

But won't merge my commits myself...

bors delegate

**Review by Saviq — APPROVED** [review]
Looks legit, and `dpkg-gensymbols` is happy.

**Comment by Saviq** [comment]
But now there's another conflict…

**Comment by AlanGriffiths** [comment]
> But now there's another conflict…

Looks trivial to fix, let me check...

**Review by Saviq — APPROVED** [review]
bors r=AlanGriffiths

**Pull request merged**